Feature/File Upload Exercise submission

maxr96 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This is the second part of implementing File Upload Exercise. It focuses on allowing the users to submit to upload a file for File Upload Exercise.
### Description
<!-- Describe your changes in detail -->
For submitting a File, student can use a simple File Upload input. He can select a file from one of the allowed file types. He can only Submit the exercise and not save. After submitting the exercise can not be changed. The File should be stored under the exercise path and be accessible under the following path: `/api/files/file-upload-exercises/{fileUploadExerciseId}/submissions/{fileUploadSubmissionId}/{filename}`, so tutor can open it for the assessment (will be implemented in the next PR). Student can also open submitted file to view it after the submission. The submitted file is stored on the server under the path `uploads\file-upload-exercise\{fileUploadExerciseId}\{fileUploadSubmissionId}`.
NOTE: This PR includes a lot of code duplication, which is noted in the issue #819 and will be resolved after the File Upload Exercise is fully implemented.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Overview
3. Select a Course that has File Upload Exercise
4. Click 'Start Exercise' on File Upload Exercise
5. Click 'Browse' to open a File Explorer and select a File you want to upload
6. Click 'Submit'
7. Now you can view the submission, but the input element is grayed out.

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
#### File Upload Exercise in Course Details
![Screenshot_2019-09-06 Course details](https://user-images.githubusercontent.com/17262376/64431321-7a8bfd80-d0ba-11e9-99c2-55975e637805.png)

#### Start Exercise
![Screenshot_2019-09-08 File Upload Exercises](https://user-images.githubusercontent.com/17262376/64481220-ffb90480-d1d7-11e9-9ff0-3e1b192023e6.png)


#### File to upload selected
![Screenshot_2019-09-08 File Upload Exercises(1)](https://user-images.githubusercontent.com/17262376/64481222-047db880-d1d8-11e9-8b57-676f96572277.png)

#### Submitted exercise with no results yet
![Screenshot_2019-09-06 Exercise details](https://user-images.githubusercontent.com/17262376/64431035-e1f57d80-d0b9-11e9-823e-1b5a6c93cd3e.png)
---
![Screenshot_2019-09-16 File Upload Exercises](https://user-images.githubusercontent.com/17262376/64949649-f2f07c80-d879-11e9-971b-50841507e411.png)

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/repository/FileUploadSubmissionRepository.java:
```diff
@@ -1,15 +1,59 @@
+     * @return the number of submissions belonging to the course id, which have the submitted flag set to true and the submission date before the exercise due date, or no exercise
+     *         due date at all
+     */
+    @Query("SELECT COUNT (DISTINCT submission) FROM FileUploadSubmission submission WHERE submission.participation.exercise.course.id = :#{#courseId} AND submission.submitted = TRUE AND (submission.submissionDate < submission.participation.exercise.dueDate OR submission.participation.exercise.dueDate IS NULL)")
```
The repository methods are copy and pasted from other repositories, e.g. ModelingSubmissionRepository.
Especially from a consistency approach this is not a good way to handle this as somebody might change one of the other repositories but not this one.

I would propose to refactor that you just add these methods into the SubmissionRepository and use the service to  to identify if the found submission has the correct class type (normally an edge case). 
This should then be streamlined in another PR with the other repositories.

maxr96 replied:
I did it for methods that return count, but I am not sure how to properly do it for ones that return `FileUploadSubmission` type.  I found out an article about JPA Query [SPEL expressions ](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#jpa.query.spel-expressions). We probably can use this idea, but then we need to change how `SubmissionRepository` is used for all services

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/service/FileUploadSubmissionService.java:
```diff
@@ -0,0 +1,325 @@
+     */
+    @Transactional
+    public FileUploadSubmission handleFileUploadSubmission(FileUploadSubmission fileUploadSubmission, FileUploadExercise fileUploadExercise, Principal principal) {
+        Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginAnyState(fileUploadExercise.getId(), principal.getName());
```
This again is a copy & pasted method from the `TextSubmissionService`. Please find a way to streamline this - we can't just duplicate the code for every new exercise type.

There are more examples in this file but I won't list every one of it here (you can check this yourself in e.g. IDEA Ultimate as duplicated code is marked in the editor).

thilo-behnke replied:
As discussed we refactor this in a later PR

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -201,6 +207,42 @@ public FileResource(FileService fileService, ResourceLoader resourceLoader, User
+        Submission submission = optionalSubmission.get();
+        try {
+            byte[] file = fileService.getFileForPath(Constants.FILE_UPLOAD_SUBMISSION_FILEPATH + submission.getId() + '/' + filename);
+            if (file == null) {
```
This is also a duplicated fragment to lines 301-309.

maxr96 replied:
Very good point, I moved the whole `try-catch` block into a separate methods. Now, additionally it should handle properly different MIME types, not only pdf

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -33,81 +46,327 @@
+    @GetMapping("/exercises/{exerciseId}/file-upload-submissions")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<List<FileUploadSubmission>> getAllFileUploadSubmissions(@PathVariable Long exerciseId, @RequestParam(defaultValue = "false") boolean submittedOnly,
+            @RequestParam(defaultValue = "false") boolean assessedByTutor) {
```
Also here there are a lot of similiarities to e.g. the `TextSubmissionResource`. Why do we duplicate these methods for every new exercise type?

thilo-behnke replied:
As discussed we refactor this in a later PR

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -33,81 +46,327 @@
+        if (!authCheckService.isAtLeastTeachingAssistantForExercise(exercise)) {
+            return forbidden();
+        }
+        if (!(exercise instanceof FileUploadExercise)) {
```
This condition is always false because fileUploadExerciseService.findOne(exerciseId) will retrieve a FileUploadExercise or null (but never another exercise type).

maxr96 replied:
Right. Deleted it

maxr96 commented on src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java:
```diff
@@ -16,11 +17,41 @@
     // jhipster-needle-entity-add-field - JHipster will add fields here, do not remove
 
+    /*
+     * NOTE: The file management is necessary to differentiate between temporary and used files and to delete used files when the corresponding exercise is deleted. The workflow is
```
I updated the doc, please, check if it is better explained now

maxr96 commented on src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java:
```diff
@@ -16,11 +17,41 @@
+        }
+    }
+
+    @PostPersist
```
Updated the class doc, if it is unclear, then I will add comment here as well

thilo-behnke replied:
:+1:

maxr96 commented on src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java:
```diff
@@ -16,11 +17,41 @@
+    @PostLoad
+    public void onLoad() {
+        // replace placeholder with actual id if necessary (this is needed because changes made in afterCreate() are not persisted)
+        if (getFilePath() != null && getFilePath().contains(Constants.FILEPATH_ID_PLACHEOLDER)) {
```
You are right!

maxr96 commented on src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java:
```diff
@@ -129,6 +129,7 @@ protected void configure(HttpSecurity http) throws Exception {
             .antMatchers("/api/account/reset-password/finish").permitAll()
             .antMatchers("/api/lti/launch/*").permitAll()
             .antMatchers("/api/files/attachments/**").permitAll()
+            .antMatchers("/api/files/file-upload-submission/**").permitAll()
```
I did analogous to lecture attachments, that also bypass our authentication. However, we can also think about better why to do it to still have authentication in place

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -201,6 +209,24 @@ public FileResource(FileService fileService, ResourceLoader resourceLoader, User
+     */
+    @GetMapping("/files/file-upload-submission/{submissionId}/{filename:.+}")
+    @PreAuthorize("permitAll()")
+    public ResponseEntity getFileUploadSubmission(@PathVariable Long submissionId, @PathVariable String filename) {
```
This implementation is missing the temporary access token: https://github.com/ls1intum/Artemis/pull/696/files#diff-32ac644447ff79fbf11c9c178d4ab71fR245

maxr96 replied:
Good suggestion, added temporary access token

thilo-behnke commented on src/main/webapp/app/file-upload-submission/file-upload-submission.component.ts:
```diff
@@ -0,0 +1,193 @@
+    private submissionConfirmationText: string;
+
+    constructor(
+        private route: ActivatedRoute,
```
Half of the services are unused.

maxr96 replied:
Yes, removed them

thilo-behnke commented on src/main/webapp/app/file-upload-submission/file-upload-submission.component.ts:
```diff
@@ -0,0 +1,193 @@
+     * @param $event {object} Event object which contains the uploaded file
+     */
+    setFileSubmissionForExercise($event: any): void {
+        if ($event.target.files.length) {
```
I think we also need to check the allowed file size. At least to avoid users to upload files larger than 10 MB. Maybe this could be a setting in future.

thilo-behnke replied:
This is already supported from jhipster. Just an important info that users can't upload more than 10MB.

thilo-behnke replied:
Another problem that we noticed here:
Try uploading a JAR or XLSX, it will be denied with the message that the file is too large. It seems like a security measure though.

maxr96 replied:
This is what we have on the File Resource: `Unsupported file type! Allowed file types: .png, .jpg, .svg, .pdf, .zip`. And it is also checked in the FileUploaderService. My suggestion is to not change it for now, only if we have a need for it. Allowing any file extension is a bad idea

thilo-behnke replied:
I think it's better now because the implementation complains that the file type is not supported (before it was always complaining about the file size).

thilo-behnke commented on src/main/webapp/app/file-upload-submission/file-upload-submission.component.ts:
```diff
@@ -0,0 +1,193 @@
+            this.erroredFile = null;
+            const fileList: FileList = $event.target.files;
+            const submissionFile = fileList[0];
+            const allowedFileExtensions = this.fileUploadExercise.filePattern.replace(/\s/g, '').split(',');
```
This should be done during serialization/deserialization. Not super important for this PR, but a Todo for the next one.

maxr96 replied:
Moved it to the `fileUploadExerciseService`

krusche commented:
I just saw a comment: File uploads should definitely be limited to 2 MB for the moment. And there should be a good error message, when the file is larger. I guess we can already check this on the client before even uploading the file to the server, but the server also has to check this (in case someone might use the REST API directly).

lovaiible commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.route.ts:
```diff
@@ -34,7 +34,7 @@ export class FileUploadExerciseResolve implements Resolve<FileUploadExercise> {
                 map((course: HttpResponse<Course>) => {
                     const fileUploadExercise = new FileUploadExercise(course.body!);
-                    fileUploadExercise.filePattern = 'pdf, png';
+                    fileUploadExercise.filePattern = '.pdf, .png';
```
Is it necessary to set a hard-coded default here for the file pattern? If not, please remove it. Also, it would be better to provide a list of file patters instead of a string.

thilo-behnke replied:
Also I wonder if we should add the dot here, it is unnecessary from a user's point of view.

maxr96 replied:
@lovaiible I would leave the hard-coded file pattern, so the user has a better idea how to enter it. Regarding the list of file patterns, this would be best implemented if we have selection of file extensions. In current state we would have to store it as a list on the server, but then handle it as a `coma-separated` string on the client.

maxr96 replied:
@thilo-behnke I think it makes more sense with a dot, this is how you usually know that it is a file extension and not a file name. But maybe I am looking on this from a different perspective

thilo-behnke replied:
To the user it should be clear that a file ending has a dot (when would it not?). This should just be something that is validated in the form.

lovaiible commented on src/main/webapp/i18n/de/fileUploadSubmission.json:
```diff
@@ -15,7 +15,13 @@
+            "confirmSubmission": "Bist du sicher, dass du abgeben möchtest? Du kannst deine Antwort nicht mehr bearbeiten, bis du eine Bewertung erhältst!",
+            "file": "Datei",
+            "fileError": "Die Datei {{ fileName }} ist zu groß. Die maximale Dateigröße beträgt 10MB",
+            "fileExtensionError": "Falsche Dateierweiterung. Bitte lädst du eine Datei mit zulässigen Dateierweiterungen hoch."
```
```suggestion
            "fileExtensionError": "Falsche Dateierweiterung. Bitte lade eine Datei mit einer zulässigen Dateierweiterung hoch."
```

ungaralex commented on src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.ts:
```diff
@@ -93,10 +94,17 @@ export class ExerciseDetailsStudentActionsComponent implements OnInit {
             const participation = this.exercise.participations[0];
             if (participation.initializationState === InitializationState.INITIALIZED || participation.initializationState === InitializationState.FINISHED) {
-                return this.exercise.type === ExerciseType.MODELING ? ParticipationStatus.MODELING_EXERCISE : ParticipationStatus.TEXT_EXERCISE;
+                return this.exercise.type === ExerciseType.MODELING
```
I think
```suggestion
                switch (this.exercise.type) {
                    case ExerciseType.MODELING: return ParticipationStatus.MODELING_EXERCISE;
                    case ExerciseType.TEXT: return ParticipationStatus.TEXT_EXERCISE;
                    default: return ParticipationStatus.FILE_UPLOAD_EXERCISE;
                }
```
is more readable.

ungaralex commented on src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts:
```diff
@@ -202,6 +209,8 @@ export class TutorExerciseDashboardComponent implements OnInit {
             submissionsObservable = this.textSubmissionService.getTextSubmissionsForExercise(this.exerciseId, { assessedByTutor: true });
         } else if (this.exercise.type === ExerciseType.MODELING) {
             submissionsObservable = this.modelingSubmissionService.getModelingSubmissionsForExercise(this.exerciseId, { assessedByTutor: true });
+        } else if (this.exercise.type === ExerciseType.FILE_UPLOAD) {
```
Please also use a switch case for the type checks in this component

maxr96 replied:
Added switch cases to type checks in this file

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -201,6 +208,29 @@ public FileResource(FileService fileService, ResourceLoader resourceLoader, User
+    @PreAuthorize("permitAll()")
+    public ResponseEntity getFileUploadSubmission(@PathVariable Long submissionId, @PathVariable String filename, @RequestParam("access_token") String temporaryAccessToken) {
+        log.debug("REST request to get file : {}", filename);
+        Optional<FileUploadSubmission> optionalSubmission = fileUploadSubmissionRepository.findById(submissionId);
```
I am not sure what can be refactored here, because the submission that I get is specific to File Upload, also later I have temporary access token validaton and I want to show custom message for them.

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/FileUploadSubmissionService.java:
```diff
@@ -0,0 +1,325 @@
+    public List<FileUploadSubmission> getFileUploadSubmissions(Long exerciseId, boolean submittedOnly) {
+        List<StudentParticipation> participations = studentParticipationRepository.findAllByExerciseIdWithEagerSubmissionsAndEagerResultsAndEagerAssessor(exerciseId);
+        List<FileUploadSubmission> submissions = new ArrayList<>();
+        for (StudentParticipation participation : participations) {
```
This is a good suggestion, but it doesn't work for me. I didn't use streams in Java much, so maybe if you can help me and correct the suggestion, I can implement that. Otherwise, let's leave it as it is for now

ungaralex replied:
There's a .stram() missing in the first line. Have you tried it after adding this?

maxr96 replied:
Yes, I definetely missed it, now it is added. Thanks for helping!

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -341,9 +346,12 @@ private void checkIfGroupsExists(Course course) {
                 numberOfSubmissions = textSubmissionService.countSubmissionsToAssessByExerciseId(exercise.getId());
             }
-            else {
+            else if (exercise instanceof ModelingExercise) {
```
Here we also want to check if exercise is instanceof of either `TextExercise`, `ModelingExercise` or `FileUploadExercise`. If we do it now in the `SubmissionService` we will receive the count for `Submissions` of all types, but we don't want to get, e.g. `ProgrammingSubmission`. I think this should be a part of next PR, as we already discussed it with Thilo. In this PR I can do `SubmissionService` more generic for different types of exercises.

maxr96 commented:
> 
> 
> From a functionality view point we are short before approving it.
> Most important issues:
> 
>     * No temp token is needed for file access (as required?)
> 
>     * The file is not shown after the submission:
> 
> 
> ![image](https://user-images.githubusercontent.com/28230611/64605697-ec1ec100-d3c4-11e9-8400-c383c6080268.png)
> 
> I also wonder if we should show the submission date here.



> 
> 
> Functionality wise, this feature looks good (except for the minor UI improvements pointed out by the other reviews.
> The main problem really is the duplicated code. It would be great if we could implement some sort of generalization for all submission types.

Thank you for the review, @ungaralex. I am aware about the code duplication and would tackle it in the next PR. I already created an issue for this. For now we need to implement it quickly and removing the code duplication will trigger changes to other exercise types as well.

maxr96 commented:
> 
> 
> From a functionality view point we are short before approving it.
> Most important issues:
> 
>     * No temp token is needed for file access (as required?)
> 
>     * The file is not shown after the submission:
> 
> 
> ![image](https://user-images.githubusercontent.com/28230611/64605697-ec1ec100-d3c4-11e9-8400-c383c6080268.png)
> 
> I also wonder if we should show the submission date here.

Good point about temp token, I added the check for it. The name of the file is shown after submission, as well as the date. 
![Screenshot_2019-09-11 File Upload Exercises](https://user-images.githubusercontent.com/17262376/64659145-7bb28700-d43a-11e9-8765-48b5ed525190.png)

maxr96 commented:
> 
> 
> I just saw a comment: File uploads should definitely be limited to 2 MB for the moment. And there should be a good error message, when the file is larger. I guess we can already check this on the client before even uploading the file to the server, but the server also has to check this (in case someone might use the REST API directly).

Should I change it in this PR? This will affect lecture attachments as well and require to change `application.yml` file

krusche commented:
> > I just saw a comment: File uploads should definitely be limited to 2 MB for the moment. And there should be a good error message, when the file is larger. I guess we can already check this on the client before even uploading the file to the server, but the server also has to check this (in case someone might use the REST API directly).
> 
> Should I change it in this PR? This will affect lecture attachments as well and require to change `application.yml` file

No this should not be the same limit as for lecture attachments. We need two different limits. The lecture attachment limit stays the same, the file upload for exercises should be limited to 2 MB. And yes please address this in this PR

maxr96 commented:
> In general it seems good, but in detail there are too many issues (just from a feature point of view).
> 
> Please pay more attention to the quality of the PR:
> 
>     * The build is failing on the second step (not even tests are executed atm). I fixed the error, but the PR should not be brought into review that rushed!
> 
>     * There are non-edge case bugs / missing features when testing the implementation (see below). I can see how this can be missed when testing manually, but especially those scenarios should be covered in the component tests.
> 
>     * A new issue was introduced that is unrelated to the PR: The user now has to add a dot to every file ending. The way the file endings are handled was not detected in the original PR, but it should have probably been some kind of auto-suggest list with a proper serialization from/to a list instead of a string. To fix this is not part of this PR, but we should not make it worse by having our users add unnecessary dots, too.
> 
> 
> **Bugs / missing implementations:**
> 
>     1. Major: Only files for the first file ending can be uploaded as the file ending field is not properly parsed:
> 
> 
> ![image](https://user-images.githubusercontent.com/28230611/64680142-804c5f00-d47d-11e9-8525-1af259bdb26e.png)
> 
> ~2) Marjo: The error message for allowed file extensions does not show which file endings are allowed~
> 
> Please ignore, the file endings are shown below the field.
> 
> ![image](https://user-images.githubusercontent.com/28230611/64680332-dde0ab80-d47d-11e9-92e7-cd7e6f1ba8db.png)
> 
>     1. Major: File extension check is case sensitive (PDF won't work if pdf is allowed).
> 
>     2. Major: There should be a changeGuard when leaving the page with a selected file that is not submitted.
> 
> 
> It's not obvious imo that you need to submit even though have already selected the file. The submit button is too far away from the upload panel. I would therefore suggest a changeGuard.
> 
>     1. Minor: When uploading the same incorrect file two times in a row, the error does not show up
> 
> 
> This is because the event for setFileSubmissionForExercise is not triggered. You could try to clear the input field on error so the browser triggers the event again.
> 
>     1. Minor: A file size that is too large should already be caught when trying to add the file, instead of letting the student submit. (unnecessary server communication)

Thank you for extensively testing the feature. I was able to fix all of the issues. I reverted the file upload exercise creation process, so the user doesn't have to specify extensions with dots. With this I also fixed removing spaces (1) and switched to case insensitive file pattern (3). This is now done before we create/update file upload submission. 
Added canDeactivate guard to address issue 4. Tested the issue 5 and it works fine now for me, the input is cleaned and user can't upload the file that is too large. I didn't change the behavior of 6 issue, but we don't have server communication anyway, since we check this on file Uploader service and reject the Promise, if file is too big.
I also addressed the request from @krusche to limit file size for file upload submission to 2 mb. We have two constants stored on the client and server side and check if the file is not too big in both places.

krusche commented:
@maxr96 Why are there still no client tests in this PR?

thilo-behnke commented on src/main/webapp/app/file-upload-submission/file-upload-submission.component.ts:
```diff
@@ -0,0 +1,207 @@
+            const fileList: FileList = $event.target.files;
+            const submissionFile = fileList[0];
+            const allowedFileExtensions = this.fileUploadExercise.filePattern.split(',');
+            if (allowedFileExtensions.some(extension => submissionFile.name.endsWith(extension))) {
```
```suggestion
            if (allowedFileExtensions.some(extension => submissionFile.name.toLowerCase().endsWith(extension))) {
```

thilo-behnke replied:
Without this the submission of test.PDF is not allowed.

maxr96 replied:
Did you create a new File Upload Exercise? Because now I lowercase the file pattern on file upload exercise creation / update

thilo-behnke replied:
Yes, you do it on the creation of the FileUploadExercise, but if the file ending of the uploaded file is PDF, the check will be `'PDF'.endsWith('pdf');`, correct?

thilo-behnke replied:
This issue is not resolved. Files with capitalized file endings can still not be uploaded:
![image](https://user-images.githubusercontent.com/28230611/64941338-a0f22b80-d866-11e9-8a42-c4717f4488e4.png)

maxr96 commented:
> Thanks for taking care of the issues! Functionality-wise there is only one fix that I commented, other than that seems fine 
> As noted by @krusche it is unclear to me why there are no client tests / why the checkbox is not ticked off. From my point of view we can also merge it in this state and do a quick follow up PR for the tests in the **next couple of days**.
> 
> Some other remarks that are more general corrections / feature requests:
> 
> * It would be nice if the student could download their submission from the view submission page (in cases like this I at least want to reassure myself that I uploaded the correct file)
> * Even though in a FileUploadExercise you allow the user to define every file ending they need for submission, the fileUploader service in the client will only allow certain types: https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/shared/http/file-uploader.service.ts#L24
> * I think in general it would be better to predefine file types in an enum in the client / persisted in the database and let the user only select those for a FileUploadExercise, similar to the categories select.

Thanks for checking it again. Regarding the client tests we can either do them in separate PR or I can work on them tomorrow and add to this PR. About other remarks:
-  This can be a good improvement for the fututure
- You probably tried not the latest version, because the file you pointed to doesn't have the new changes. There I only check the type of the file, if it is not submission file. Because if it is a submission file we check the file type on input already.
- Yes this can be a good idea. We need to discuss then what files can we allow for the users to specify.

krusche commented:
Where exactly are you storing the files on the server? What happens if two users upload files with the same file name? I think the path should include the exercise id and the user id or the participation id to prevent problems with duplicated files. Is this the case?

Also I agree to @thilo-behnke that users should see what and when they have submitted a file in the submission page and they should be able to download this file!

This is a necessary implementation before we can merge

maxr96 commented:
> Where exactly are you storing the files on the server? What happens if two users upload files with the same file name? I think the path should include the exercise id and the user id or the participation id to prevent problems with duplicated files. Is this the case?
> 
> Also I agree to @thilo-behnke that users should see what and when they have submitted a file in the submission page and they should be able to download this file!
> 
> This is a necessary implementation before we can merge
 
In current implementation files are stored under the route `/file/file-upload-submission/{file-upload-submission-id}/{filename} . I assume the submission id should be unique for each new file upload submission. But I can also include exercise id.
I already added the date and name of the file to the submission page. Will also add the possibility to download it.

krusche commented:
I think it would be easier if you create subfolders based on exercise ids and then use the file submission id. What happens if a user uploads a second file? Will the first one be deleted?
1) in case they have the same filename
2) in case they have different filenames

maxr96 commented:
> 
> 
> I think it would be easier if you create subfolders based on exercise ids and then use the file submission id. What happens if a user uploads a second file? Will the first one be deleted?
> 
>     1. in case they have the same filename
> 
>     2. in case they have different filenames

We can create a subfolder for exercise and then have subfolders there for each submission. The user is not allowed to upload a second file, once he submitted the first file he is not allowed to change this.

krusche commented:
So there is no save button as for other exercise types? Why not?

maxr96 commented:
> 
> 
> So there is no save button as for other exercise types? Why not?

That was initial requirment from you. I added this also to the description, that user can only submit and not save the exercise.

krusche commented:
Ok, then let’s leave this for now, but this has to be improved later on

thilo-behnke commented:
I can't approve the PR yet.

**Open issues:**
- Bug: File ending check is case sensitive: https://github.com/ls1intum/Artemis/pull/805#discussion_r324541393
- Bug: I get a 403 when clicking on the file link after submission
- Design: Imo the file download page looks weird, we should at least show the file name and the submission datetime next to it. (E.g. 'solution.pdf, submitted on XX. **download**').
![image](https://user-images.githubusercontent.com/28230611/64941527-43121380-d867-11e9-904f-32671b1cc6d6.png)

**Also from a PR point of view:**
- Please make sure that the build is successful before requesting the review.
- In PRs like this it gets hard to understand what has changed, you could e.g. provide a short comment on what has changed since the last review. E.g. it is not clear if you have changed something about the folder structure as suggested by @krusche.

maxr96 commented:
> 
> 
> I can't approve the PR yet.
> 
> **Open issues:**
> 
>     * Bug: File ending check is case sensitive: [#805 (comment)](https://github.com/ls1intum/Artemis/pull/805#discussion_r324541393)
> 
>     * Bug: I get a 403 when clicking on the file link after submission
> 
>     * Design: Imo the file download page looks weird, we should at least show the file name and the submission datetime next to it. (E.g. 'solution.pdf, submitted on XX. **download**').
>       ![image](https://user-images.githubusercontent.com/28230611/64941527-43121380-d867-11e9-904f-32671b1cc6d6.png)
> 
> 
> **Also from a PR point of view:**
> 
>     * Please make sure that the build is successful before requesting the review.
> 
>     * In PRs like this it gets hard to understand what has changed, you could e.g. provide a short comment on what has changed since the last review. E.g. it is not clear if you have changed something about the folder structure as suggested by @krusche.



> 
> 
> I can't approve the PR yet.
> 
> **Open issues:**
> 
>     * Bug: File ending check is case sensitive: [#805 (comment)](https://github.com/ls1intum/Artemis/pull/805#discussion_r324541393)
> 
>     * Bug: I get a 403 when clicking on the file link after submission
> 
>     * Design: Imo the file download page looks weird, we should at least show the file name and the submission datetime next to it. (E.g. 'solution.pdf, submitted on XX. **download**').
>       ![image](https://user-images.githubusercontent.com/28230611/64941527-43121380-d867-11e9-904f-32671b1cc6d6.png)
> 
> 
> **Also from a PR point of view:**
> 
>     * Please make sure that the build is successful before requesting the review.
> 
>     * In PRs like this it gets hard to understand what has changed, you could e.g. provide a short comment on what has changed since the last review. E.g. it is not clear if you have changed something about the folder structure as suggested by @krusche.

I checked the bugs that you pointed out, but all works fine for me. Exercise is saved with lowercase `filePattern` even if I enter it uppercase and I don't get 403 on submitting. Can you maybe do a screenshot of the error? Also did you try to create a new exercise after the changes?

I moved submitted time as you suggested and added file name:
![Screenshot_2019-09-16 File Upload Exercises](https://user-images.githubusercontent.com/17262376/64947009-68594e80-d874-11e9-9f34-f81f03d0a769.png)

Regarding the failing build, I just found out that the last test doesn't work when all tests are run together, but works properly separately. I will investigate more into this issue,

You are right about the comment, I only updated the description, but it would be better to add an explicit comment. From the UI perspective the main thing that I add is ability to download file by a student after submission. On the server it was mainly some fixes and refactorings. Also added client tests, as they were missing. 

Yet I didn't change the folder structure. In my opinion, it introduces some overhead, since we will need to get participation from submission to get exercise id. I think having only `submissionId` in this case is enough.

maxr96 commented:
> 
> 
> I can't approve the PR yet.
> 
> **Open issues:**
> 
>     * Bug: File ending check is case sensitive: [#805 (comment)](https://github.com/ls1intum/Artemis/pull/805#discussion_r324541393)
> 
>     * Bug: I get a 403 when clicking on the file link after submission
> 
>     * Design: Imo the file download page looks weird, we should at least show the file name and the submission datetime next to it. (E.g. 'solution.pdf, submitted on XX. **download**').
>       ![image](https://user-images.githubusercontent.com/28230611/64941527-43121380-d867-11e9-904f-32671b1cc6d6.png)
> 
> 
> **Also from a PR point of view:**
> 
>     * Please make sure that the build is successful before requesting the review.
> 
>     * In PRs like this it gets hard to understand what has changed, you could e.g. provide a short comment on what has changed since the last review. E.g. it is not clear if you have changed something about the folder structure as suggested by @krusche.



> 
> 
> I can't approve the PR yet.
> 
> **Open issues:**
> 
>     * Bug: File ending check is case sensitive: [#805 (comment)](https://github.com/ls1intum/Artemis/pull/805#discussion_r324541393)
> 
>     * Bug: I get a 403 when clicking on the file link after submission
> 
>     * Design: Imo the file download page looks weird, we should at least show the file name and the submission datetime next to it. (E.g. 'solution.pdf, submitted on XX. **download**').
>       ![image](https://user-images.githubusercontent.com/28230611/64941527-43121380-d867-11e9-904f-32671b1cc6d6.png)
> 
> 
> **Also from a PR point of view:**
> 
>     * Please make sure that the build is successful before requesting the review.
> 
>     * In PRs like this it gets hard to understand what has changed, you could e.g. provide a short comment on what has changed since the last review. E.g. it is not clear if you have changed something about the folder structure as suggested by @krusche.

I checked the bugs that you pointed out, but all works fine for me. Exercise is saved with lowercase `filePattern` even if I enter it uppercase and I don't get 403 on submitting. Can you maybe do a screenshot of the error? Also did you try to create a new exercise after the changes?

I moved submitted time as you suggested and added file name:
![Screenshot_2019-09-16 File Upload Exercises](https://user-images.githubusercontent.com/17262376/64947009-68594e80-d874-11e9-9f34-f81f03d0a769.png)

Regarding the failing build, I just found out that the last test doesn't work when all tests are run together, but works properly separately. I will investigate more into this issue,

You are right about the comment, I only updated the description, but it would be better to add an explicit comment. From the UI perspective the main thing that I add is ability to download file by a student after submission. On the server it was mainly some fixes and refactorings. Also added client tests, as they were missing. 

Yet I didn't change the folder structure. In my opinion, it introduces some overhead, since we will need to get participation from submission to get exercise id. I think having only `submissionId` in this case is enough.

krusche commented:
Please use the exercise id for the folder path as requested. This is important for the administration of the server. The overhead is not too much because you have to check the permissions anyway and this involves loading the participation. The exercise is most likely already available in the hibernate cache. 

Also please make finally sure that all work in this PR is finished and the quality is good enough so that we can merge this. We are already way beyond the schedule and the assessment part is still missing 😔

maxr96 commented:
Updates since last code review:
- Moved file upload submission file to the `uploads\file-upload-exercise\{fileUploadExerciseId}\{fileUploadSubmissionId}` path on the server. The user now navigates to `/api/files/file-upload-exercises/{fileUploadExerciseId}/submissions/{fileUploadSubmissionId}/{filename}`. 
- Proper file extension is shown always, even if file have dots in the name
- Some file types were not properly converted into MIME types by `FileNameMap,getContentTypeFor(filename)`. Eventually, in a lot of the cases now it will return `application/octet-stream` which will show download dialog to the user. 
- Client tests were still strange for me. The last test was not getting input tag from template if run with other tests, but worked fine if ran separetely. I had to remove the part of checking the input tag since I couldn't find any workaround. This fixed the tests.

thilo-behnke commented on src/test/javascript/spec/mocks/mock-file-upload-submission.service.ts:
```diff
@@ -0,0 +1,22 @@
+import { EntityResponseType, FileUploadSubmission } from 'app/entities/file-upload-submission';
+import { Result } from 'app/entities/result';
+
+export const fileUploadParticipation = new StudentParticipation();
```
The tests are fixed, but don't keep the objects in the scope of the MockService.
It would be better to e.g. provide a small factory to guarantee that a new object is created.
Otherwise this will likely cause issues again in the future.

E.g.

```
export const createFileUploadSubmission = () => {
  const fileUploadSubmission = new FileUploadSubmission();
  fileUploadSubmission.submitted = false;
  fileUploadSubmission.participation = fileUploadParticipation;
  fileUploadSubmission.id = 1;
  return fileUploadSubmission;
}
```

maxr96 replied:
Thank you for pointing this out. I didn't understand how it should work at first. Now it is changed

lovaiible commented on src/main/webapp/app/file-upload-submission/file-upload-submission.component.ts:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    /**
+     * Returns false if user selected a file, but didn't submit the exercise, false otherwise.
```
The documentation doesn't make sense atm, because you say that either case `false` is returned. Also I am a bit confused: do we need `Observable<boolean>` as a return type if we only return `boolean`?

maxr96 replied:
Good catch. That was a typo and returning `boolean` is cleaner. Changed.

maxr96 commented on src/test/javascript/spec/component/file-upload-submission/file-upload-submission.spec.ts:
```diff
@@ -0,0 +1,171 @@
+    it('Incorrect file type can not be submitted', fakeAsync(() => {
+        // Ignore console errors and window confirm
+        console.error = jest.fn();
+        window.confirm = () => {
```
In this case, I don't need to do this since for each test it will load with original functionality and can't be overridden by other test. Also realized that setting `window.confirm` is redundant in this line

maxr96 commented:
> Good job implementing all the feedbacks! I have added some comments here and there, maybe you can have a look at them.
> One thing that I noticed while testing is that the label is still outdated. You don't define extensions with a leading period character anymore, do you?
> ![image](https://user-images.githubusercontent.com/33764166/65046128-4b01ae80-d960-11e9-8278-e8e33bfaf2c4.png)

Thank you for the review @lovaiible. Good catch with file pattern tooltip. Yes it should not have dots, I forgot to update the text after removing the requirement for the dots. I also resolved other comments.

krusche commented:
@maxr96  I optimized imports on `develop`. Please resolve the conflicts

krusche commented:
@maxr96 Is everything finished now? I still see some unresolved conversations.

maxr96 commented:
> 
> 
> @maxr96 Is everything finished now? I still see some unresolved conversations.

@krusche Yes all the comments are resolved now. I merged the optimized imports.

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java:
```diff
@@ -3,24 +3,48 @@
+        if (filePath != null) {
+            // delete old file if necessary
+            String[] path = filePath.split("/");
+            fileService.manageFilesForUpdatedFilePath(filePath, null, FileUploadSubmission.buildFilePath(Long.parseLong(path[4]), getId()), getId(), true);
```
this looks like code that can easily fail, how do you make sure that the path actually has 4 entries? Please write safer code!

ungaralex replied:
I think this should also really not depend on the exerciseId in the filePath itself, but rather just give the whole path to a delete method in the FileService. Then again, the actual path right now in the database is not the path in the filesystem, but the API path e.g. "/api/files/file-upload-exercises/30/submissions/100/some_file.jar". I think this should be something like "uploads/file-upload-exercises/30/submissions/100/some_file.jar"

ungaralex replied:
For now, I've added some checks and variables. But I really think we should remove the tight coupling between URLs and getting IDs for exercises, submissions, etc. in general.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/FileService.java:
```diff
@@ -144,6 +145,10 @@ public String actualPathForPublicPath(String publicPath) {
             return Constants.LECTURE_ATTACHMENT_FILEPATH + lectureId + filename;
         }
+        if (publicPath.contains("files/file-upload-exercises")) {
+            String[] subpath = publicPath.replace(filename, "").replace("/api/files/file-upload-exercises/", "").split("/");
```
This is bad code. I think you try to parse some ids here, then please also use variable names that include the ids like above `lectureId`. Also make sure that this code cannot fail by checking if these elements are actually available

ungaralex replied:
Added explicit number check and used new variables

krusche commented on src/main/java/de/tum/in/www1/artemis/service/FileService.java:
```diff
@@ -179,6 +184,10 @@ private String publicPathForActualPath(String actualPath, Long entityId) {
         }
+        if (actualPath.contains(Constants.FILE_UPLOAD_EXERCISES_FILEPATH)) {
+            Path path = Paths.get(actualPath);
+            return "/api/files/file-upload-exercises/" + path.getName(2) + "/submissions/" + id + "/" + filename;
```
what is path.getName(2) --> this is bad code. Put it into a variable

ungaralex replied:
Added explicit number check and put it into a variable

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -53,38 +60,51 @@
         // NOTE: Maximum file size is set in resources/config/application.yml
         // Currently set to 10 MB
 
+        if (isSubmission && file.getSize() > 2_000_000) {
```
Please move this to a constant and make sure to write it as `2 * 1024 * 1024`.

ungaralex replied:
Moved it

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -53,38 +60,51 @@
     @PostMapping("/fileUpload")
-    @PreAuthorize("hasAnyRole('ADMIN', 'INSTRUCTOR', 'TA')")
-    public ResponseEntity<String> saveFile(@RequestParam(value = "file") MultipartFile file, @RequestParam("keepFileName") Boolean keepFileName) throws URISyntaxException {
+    @PreAuthorize("hasAnyRole('ADMIN', 'INSTRUCTOR', 'TA', 'USER')")
```
This creates a problem, because now any user can upload a file. I think we need to add additional checks now, otherwise this endpoint becomes unsecure!

ungaralex replied:
Actually, I think we nee a dedicated endpoint for uploading and creating temporary files for the upload exercise. Students could bombard the server with files for the same exercise by manually calling the endpoint. 
They could even just overwrite other submission in theory if they are using filenames chosen by other users. The temporary file would overwrite the actual submission before the other user calls the saveExercise endpoint

krusche replied:
This sounds much better in terms of security. However we should try to realize this with the most pragmatic approach. It might e.g. be difficult to block a user. But I’m curious how you would implement this 😉

ungaralex replied:
So right now, there is a check for the following properties:

- File size
- Is the course the same as for the exerciseId?
- Check whtether the user already uploaded

As long as those 3 don't pass, nothing will be saved.

krusche replied:
please remove 'USER' in the pre authorize. I guess we could even remove 'TA' here, but I am not sure. Please double check if 'TA' is necessary

maxr96 replied:
This endpoint is used for lecture attachments, uploading course icon, quiz exercise generation, editing questions for quiz exercise. At least the last one is accessible for tutor from the client

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -53,38 +60,51 @@
-        if (!fileExtension.equalsIgnoreCase("png") && !fileExtension.equalsIgnoreCase("jpg") && !fileExtension.equalsIgnoreCase("jpeg") && !fileExtension.equalsIgnoreCase("svg")
-                && !fileExtension.equalsIgnoreCase("pdf") && !fileExtension.equalsIgnoreCase("zip")) {
+        List<String> allowedFileExtensions = Arrays.asList("png", "jpg", "zip", "pdf", "svg", "jpeg");
+        if (!isSubmission && !allowedFileExtensions.stream().anyMatch(fileExtension::equalsIgnoreCase)) {
```
You do not check if the file submitted by a student has the correct file ending as specified in the corresponding FileUploadExercise.filePattern. 

However, this is important, otherwise students could upload any file using this endpoint!

ungaralex replied:
Added a check on the server in the new endpoint

krusche commented on src/main/webapp/app/file-upload-submission/file-upload-submission.module.ts:
```diff
@@ -0,0 +1,27 @@
+    providers: [{ provide: JhiLanguageService, useClass: JhiLanguageService }],
+})
+export class ArtemisFileUploadSubmissionModule {
+    constructor(private languageService: JhiLanguageService, private languageHelper: JhiLanguageHelper) {
```
Please remove this. In the last `develop` update, we have removed this from all angular modules

maxr96 replied:
Removed

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -457,9 +466,12 @@ private void checkIfGroupsExists(Course course) {
+            else if (exercise instanceof ModelingExercise) {
                 numberOfParticipations += modelingSubmissionService.countSubmissionsToAssessByExerciseId(exercise.getId());
             }
+            else {
```
please add a check for FileUploadExercise here

krusche commented on src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java:
```diff
@@ -129,6 +129,7 @@ protected void configure(HttpSecurity http) throws Exception {
             .antMatchers("/api/account/reset-password/finish").permitAll()
             .antMatchers("/api/lti/launch/*").permitAll()
             .antMatchers("/api/files/attachments/**").permitAll()
+            .antMatchers("/api/files/file-upload-submission/**").permitAll()
```
after the last changes we do not need this anymore, right?

maxr96 replied:
Yes this is still needed for file download

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -343,9 +348,12 @@ private void checkIfGroupsExists(Course course) {
+            else if (exercise instanceof ModelingExercise) {
                 numberOfSubmissions += modelingSubmissionService.countSubmissionsToAssessByExerciseId(exercise.getId());
             }
+            else {
```
```suggestion
            else if (exercise instanceof FileUploadExercise) {
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -201,6 +218,32 @@ public FileResource(FileService fileService, ResourceLoader resourceLoader, User
+            return ResponseEntity.badRequest().build();
+        }
+        if (!validateTemporaryAccessToken(temporaryAccessToken, filename)) {
+            String errorMessage = "You don't have the access rights for this file! Please login to Artemis and download the file in the corresponding File Upload Submission";
```
```suggestion
            String errorMessage = "You don't have the access rights for this file! Please login to Artemis and download the file in the corresponding exercise";
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:
```diff
@@ -248,38 +291,59 @@ public ResponseEntity getLectureAttachment(@PathVariable Long lectureId, @PathVa
-            if (filename.endsWith(".pdf")) {
-                mediaType = "application/pdf";
+            FileNameMap fileNameMap = URLConnection.getFileNameMap();
+            String mimeType = fileNameMap.getContentTypeFor(filename);
```
Did you test that this works for pdf and zip files for lectures as well?

maxr96 replied:
Yes this works

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -33,81 +45,314 @@
-        FileUploadSubmission result = fileUploadSubmissionRepository.save(fileUploadSubmission);
-        return ResponseEntity.ok().headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, fileUploadSubmission.getId().toString())).body(result);
+    @GetMapping("/file-upload-submissions/{submissionId}")
+    public ResponseEntity<FileUploadSubmission> getFileUploadSubmission(@PathVariable Long submissionId) {
```
`@PreAuthorize("hasAnyRole('USER','TA','INSTRUCTOR','ADMIN')")` is missing

krusche replied:
or `@PreAuthorize("hasAnyRole('TA','INSTRUCTOR','ADMIN')")` in case this should only be used by tutors and above

maxr96 replied:
Add `@PreAuthorize("hasAnyRole('TA','INSTRUCTOR','ADMIN')")`. User should not be able to acess it

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -33,81 +45,314 @@
 
     /**
-     * PUT /file-upload-submissions : Updates an existing fileUploadSubmission.
+     * GET /file-upload-submissions/:id : get the "id" fileUploadSubmission.
```
A documentation is missing for what we need this API call.
Is it for tutors?

maxr96 replied:
Yes, it is for tutors assessment, changed the description

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -33,81 +45,314 @@
-    public List<FileUploadSubmission> getAllFileUploadSubmissions() {
-        log.debug("REST request to get all FileUploadSubmissions");
-        return fileUploadSubmissionRepository.findAll();
+    @GetMapping("/exercises/{exerciseId}/file-upload-submissions")
```
for which use case do we need this API endpoint?

maxr96 replied:
This is used to list submission on the exercise dashboard to the assessor (started and completed)

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -33,81 +45,314 @@
-        log.debug("REST request to delete FileUploadSubmission : {}", id);
-        fileUploadSubmissionRepository.deleteById(id);
-        return ResponseEntity.ok().headers(HeaderUtil.createEntityDeletionAlert(applicationName, true, ENTITY_NAME, id.toString())).build();
+    @DeleteMapping("/file-upload-submissions/{submissionId}")
```
Seems that everyone can invoke this API endpoint, which is **really really bad**!!!
Make sure that only instructors of the course and admins can invoke this and make sure to `log.info(..)` who has executed this call so that it will be visible in the logs on the test / production server

maxr96 replied:
This endpoint is not needed, as we remove submission only when removing the exercise. Removed

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -33,81 +45,314 @@
+        }
+        else {
+            return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(applicationName, true, "fileUploadExercise", "wrongExerciseType",
+                    "The exercise of the participation is not a modeling exercise.")).body(null);
```
modeling exercise? 👎

maxr96 replied:
Changed

krusche commented on src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.ts:
```diff
@@ -93,10 +94,20 @@ export class ExerciseDetailsStudentActionsComponent implements OnInit {
+                        return ParticipationStatus.MODELING_EXERCISE;
+                    case ExerciseType.TEXT:
+                        return ParticipationStatus.TEXT_EXERCISE;
+                    default:
```
please use a proper case for this in all code places, because we have more exercise types and there might be even more in the future!

maxr96 replied:
Changed this. Didn't find any more cases like this

ungaralex commented:
The access token concept in general might be insecure. Users can login and use the API to get an access token for any file they want since we don't make a distinction between "public" files and submissions in the `getTemporaryFileAccessToken` method. 
They could then just get a token for the submission of another student and download his solution, or did I overlook something? @maxr96 @krusche

maxr96 commented:
> 
> 
> The access token concept in general might be insecure. Users can login and use the API to get an access token for any file they want since we don't make a distinction between "public" files and submissions in the `getTemporaryFileAccessToken` method.
> They could then just get a token for the submission of another student and download his solution, or did I overlook something? @maxr96 @krusche

I believe that this is not the case. I checked the discussion on the respecrive PR #696. Well the request to create a token can only be created by a client and it is bound to specific file and user name.

krusche commented on src/main/webapp/i18n/de/fileUploadSubmission.json:
```diff
@@ -15,7 +15,17 @@
+            "file": "Datei",
+            "fileTooBigError": "Die Datei {{ fileName }} ist zu groß. Die maximale Dateigröße beträgt 2 MB",
+            "fileExtensionError": "Falsche Dateierweiterung. Bitte lade eine Datei mit einer der unten aufgeführten zulässigen Dateierweiterungen hoch.",
+            "fileUploadError": "Deine Einreichung zur Datei {{ fileName }} ist fehlgeschlagen!",
```
```suggestion
            "fileUploadError": "Deine Einreichung der Datei {{ fileName }} ist fehlgeschlagen!",
```
